### ISSUE TITLE
Explore splitting foundation to a separate package without strict dependency on dart:ui

### ISSUE THREAD

**Issue by dnfield** [flutter/flutter#29230]
See https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/foundation/README.md

We have a customer request to consume parts of Flutter's foundation pacage in a Dart isolate that isn't using Flutter - specifically to use ValueNotifier/ChangeNotifier.

Today there are only a small handful of dependencies on dart:ui in foundation, and we already know we want to get rid of them.  If we moved foundation to another package (in the flutter/flutter repo), we may be able to make dart:ui a conditional import so that we could define VoidCallback, hashValues, hashList, and lerpDouble pass through to dart:ui if it's available or define them in the foundation package otherwise.  The rest of flutter/flutter could take whatever is exported from the foundation package rather than from dart:ui

/cc @mehmetf @goderbauer @tvolkert @Hixie

**Comment by jonahwilliams** [comment]
ValueNotifier/ChangeNotifier are trivial classes, what is the problem with re-creating them externally? Likewise, the entirety of flutter assumes dart:ui is present, and we really don't have the analyzer capabilities to ensure we write a program which can run without it.

**Comment by dnfield** [comment]
They depend on `VoidCallback`, which is defined in dart:ui.

**Comment by dnfield** [comment]
I'm not convinced this has to be an API break.  We could re-export everything transparently.

**Comment by mehmetf** [comment]
@dnfield Is it possible to be more surgical about this? I don't think the Notifier classes depend on dart:ui. Perhaps we can start by moving them out of foundation? I think that might be a better path to eventually emptying out foundation; it does not have to move all at once.

**Comment by jonahwilliams** [comment]
VoidCallback is a typedef, these are structurally typed.

**Comment by jonahwilliams** [comment]
```
typedef VoidCallback = void Function();
typedef VoidCallback2 = void Function();

void main() {
  VoidCallback x = () {};
  VoidCallback2 y = x; // OK
}

```

**Comment by dnfield** [comment]
As another option, what if we looked to move the Notifier classes to the Dart SDK?  /cc @a-siva @zra

**Comment by zanderso** [comment]
Can I get a link to the Notifier classes? =)

**Comment by dnfield** [comment]
https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/foundation/change_notifier.dart

**Comment by jonahwilliams** [comment]
Some more context:
In google3 there are additional platform restrictions. Whether or not the individual library uses dart:ui doesn't matter, the blaze target does and that restricts the use of foundation.

The only way to make this usable would be to split it into a separate blaze target, and likely a separate flutter package.

**Comment by tvolkert** [comment]
I'd like to push more on https://github.com/dart-lang/sdk/issues/27791 and https://github.com/dart-lang/sdk/issues/25217, which, if they were done, would let people use foundation out of the box.

**Comment by hterkelsen** [comment]
I support this, but I think you should also factor out the platform channel code (in src/services) along with the foundation code.

**Comment by Hixie** [comment]
I concur with Todd's comment above.
I would rather we didn't export lots of packages. It makes everything more difficult to reason about.

**Comment by jxson** [comment]
There has been some discussion about the design of some of Fuchsia's Dart packages which would like to use Flutter's foundation package (primarily for the ValueChanged, ValueListenable, ChangeNotifier classes) but the transitive dep on dart:ui is problematic.

We have a workaround so this isn't blocking, just curious what the status of this is...

**Comment by jacob314** [comment]
#32464 is a great step towards making this possible by enabling the foundation code to run on the web. This CL may be sufficient or we may need a follow on CL that enables flutter/foundation to work when both dart:ui is and dart:html are not present (e.g. command line).

Strawman implementation:
* Write a script which we can put in the dev/bots directory, that takes the foundation directory, massages it into the shape of a package that imports flutter/foundation if it's available.
* Creates a pubspec.yaml with a version matching the current flutter version (from the /version file probably)
* Publishes it
* Call that from the LUCI infrastructure.

**Comment by filiph** [comment]
With people increasingly using `ChangeNotifier` for their view models and other `flutter:foundation` classes (like `Color`) in their business logic, it will be progressively more helpful to have that extracted as a minimal package that doesn't depend on Flutter SDK.

For this reason, https://github.com/dart-lang/sdk/issues/27791 and https://github.com/dart-lang/sdk/issues/25217 won't help by themselves — I don't think the Dart team will ever agree to have `ChangeNotifier` as a `dart:core` class.

**Comment by tvolkert** [comment]
There are two related but distinct topics that I've seen discussed here and in related bugs:

1. Removing the foundation library's dependency on `dart:ui` (dart-lang/sdk#27791 and dart-lang/sdk#25217)
1. Allowing developers to use foundation classes outside the context of a flutter SDK (e.g. for server-side Dart).

**Comment by Hixie** [comment]
(Note that `Color` isn't from `foundation`, it's from `dart:ui`.)

**Comment by rrousselGit** [comment]
I agree with the statement that ValueNotifier/ChangeNotifier are trivial

What's less trivial is DiagnosticsProperty and its family. They are useful to make things readable inside the devtool.
But forking their sources simply doesn't work – we lose the ability to "expand" nodes and it behaves like a custom `toString`.

**Comment by rrousselGit** [comment]
As a comeback to this – what can I do to get things done on this issue

I don't mind making a PR to extract `diagnostrics` on `flutter/packages`. But I'm not sure if that's enough or if there's anything else blocking?

**Comment by timosturm** [comment]
Are there any updates on this? I often find myself having problems because of this transit dependency, as it breaks flutters driver tests. It also leads to unorganized code because every functionality depending on 'foundations.dart' must then be in a different file.
